File: Homework/Objective-C/HW23/HW23/ViewController.m
//
//  ViewController.m
//  HW23
//
//  Created by Руслан on 18.10.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <UIGestureRecognizerDelegate>

@property (weak, nonatomic) IBOutlet UIImageView *ball;
@property (assign, nonatomic) CGFloat ballScale;
@property (assign, nonatomic) CGFloat ballRotation;

#define pi 3.14
#define minus_pi -3.14

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // tap gesture
    UITapGestureRecognizer *tapGesture =
    [[UITapGestureRecognizer alloc] initWithTarget:self
                                            action:@selector(handleTap:)];
    [self.view addGestureRecognizer:tapGesture];
    
    
    // right swipe gesture
    UISwipeGestureRecognizer *swipeRightGesture =
    [[UISwipeGestureRecognizer alloc] initWithTarget:self
                                              action:@selector(handleSwipeRight:)];
    
    swipeRightGesture.direction = UISwipeGestureRecognizerDirectionRight;
    [self.view addGestureRecognizer:swipeRightGesture];
    
    // left swipe gesture
    UISwipeGestureRecognizer *swipeLeftGesture =
    [[UISwipeGestureRecognizer alloc] initWithTarget:self
                                              action:@selector(handleSwipeLeft:)];
    
    swipeLeftGesture.direction = UISwipeGestureRecognizerDirectionLeft;
    [self.view addGestureRecognizer:swipeLeftGesture];
    
    // double tap gesture
    UITapGestureRecognizer *tapDoubleGesture =
    [[UITapGestureRecognizer alloc] initWithTarget:self
                                            action:@selector(handleTapDouble:)];
    tapDoubleGesture.numberOfTapsRequired = 2;
    [self.view addGestureRecognizer:tapDoubleGesture];
    
    [tapGesture requireGestureRecognizerToFail:tapDoubleGesture];
    
    
    // pinch gesture
    UIPinchGestureRecognizer *pinchGesture =
    [[UIPinchGestureRecognizer alloc] initWithTarget:self
                                              action:@selector(handlePinch:)];
    pinchGesture.delegate = self;
    
    [self.view addGestureRecognizer:pinchGesture];
    
    // rotation gesture
    UIRotationGestureRecognizer *rotationGesture =
    [[UIRotationGestureRecognizer alloc] initWithTarget:self
                                                 action:@selector(handleRotation:)];
    rotationGesture.delegate = self;
    
    [self.view addGestureRecognizer:rotationGesture];
}

#pragma mark - Helpful functions

- (void) rotateView:(UIView *)view withAngle:(CGFloat)angle
         andAnimationDuration:(NSTimeInterval)animationDuration usingFullCircle:(BOOL)fullCircle {
    
    [UIView animateWithDuration:animationDuration
                          delay:0
                        options:UIViewAnimationOptionCurveLinear
                     animations:^{
                         
                         self.ball.transform = CGAffineTransformRotate(self.ball.transform, angle);
                     }
                     completion:^(BOOL finished) {
                         if (fullCircle) {
                             [self rotateView:self.ball withAngle:angle
                                   andAnimationDuration:animationDuration usingFullCircle:NO];
                         }
                     }];
}

#pragma mark - Touches

- (void) handleTap:(UITapGestureRecognizer *)tapGesture {
    
    [UIView animateWithDuration:2
                          delay:0
                        options:UIViewAnimationOptionAllowUserInteraction | UIViewAnimationOptionBeginFromCurrentState
                     animations:^{
                         self.ball.center = [tapGesture locationInView:self.view];
                     }
                     completion:^(BOOL finished) {
                     }];
}

- (void) handleSwipeRight:(UISwipeGestureRecognizer *)swipeRightGesture {
    
    NSLog(@"handleSwipeRight");
    
    [self rotateView:self.ball withAngle:pi andAnimationDuration:0.5f usingFullCircle:YES];
}

- (void) handleSwipeLeft:(UISwipeGestureRecognizer *)swipeLeftGesture {
    
    NSLog(@"handleSwipeLeft");
    
    [self rotateView:self.ball withAngle:minus_pi andAnimationDuration:0.5f usingFullCircle:YES];
}

- (void) handleTapDouble:(UITapGestureRecognizer *)tapDoubleGesture {
    NSLog(@"handleTapDouble");
    [self.ball.layer removeAllAnimations];
}

- (void) handlePinch:(UIPinchGestureRecognizer *)pinchGesture {
    
    if (pinchGesture.state == UIGestureRecognizerStateBegan) {
        self.ballScale = 1.f;
    }
    
    CGFloat scale = 1.f + pinchGesture.scale - self.ballScale;
    
    self.ball.transform = CGAffineTransformScale(self.ball.transform, scale, scale);
    
    self.ballScale = pinchGesture.scale;
}

- (void) handleRotation:(UIRotationGestureRecognizer *)rotationGesture {
    
    if (rotationGesture.state == UIGestureRecognizerStateBegan) {
        self.ballRotation = 0.f;
    }
    
    CGFloat rotation = rotationGesture.rotation - self.ballRotation;
    
    self.ball.transform = CGAffineTransformRotate(self.ball.transform, rotation);
    
    self.ballRotation = rotationGesture.rotation;
}

#pragma mark - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer
        shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    return YES;
}

@end

File: Homework/Swift/HW23/HW23/ViewController.swift
//
//  ViewController.swift
//  HW23
//
//  Created by Руслан on 18.10.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UIGestureRecognizerDelegate {

    @IBOutlet weak var ball: UIImageView!
    let pi: CGFloat = 3.14
    let minus_pi: CGFloat = -3.14
    var ballScale: CGFloat!
    var ballRotation: CGFloat!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // tap gesture
        let tapGesture = UITapGestureRecognizer.init(target: self, action: #selector(handleTap))
        self.view.addGestureRecognizer(tapGesture)
        
        
        // right swipe gesture
        let rightSwipeGesture = UISwipeGestureRecognizer.init(target: self, action: #selector(handleRightSwipe))
        rightSwipeGesture.direction = .right
        self.view.addGestureRecognizer(rightSwipeGesture)
        
        // left swipe gesture
        let leftSwipeGesture = UISwipeGestureRecognizer.init(target: self, action: #selector(handleLeftSwipe))
        leftSwipeGesture.direction = .left
        self.view.addGestureRecognizer(leftSwipeGesture)
        
        
        // double tap gesture
        let doubleTapGesture = UITapGestureRecognizer.init(target: self, action: #selector(handleDoubleTap))
        doubleTapGesture.numberOfTapsRequired = 2
        self.view.addGestureRecognizer(doubleTapGesture)
        
        tapGesture.require(toFail: doubleTapGesture)
        
        
        // pinch gesture
        let pinchGesture = UIPinchGestureRecognizer.init(target: self, action: #selector(handlePinch))
        pinchGesture.delegate = self
        self.view.addGestureRecognizer(pinchGesture)
        
        // rotation gesture
        let rotationGesture = UIRotationGestureRecognizer.init(target: self, action: #selector(handleRotation))
        rotationGesture.delegate = self
        self.view.addGestureRecognizer(rotationGesture)
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
// MARK: - Helpful functions
    
    func rotateView(_ view: UIView, withAngle angle: CGFloat,
                    andAnimationDuration animationDuration: TimeInterval, usingFullCircle fullCircle: Bool) {
        
        UIView.animate(withDuration: animationDuration,
                       delay: 0,
                       options: .curveLinear,
                       animations: {
                        view.transform = view.transform.rotated(by: angle)
            }) { (true) in
                if fullCircle == true {
                    self.rotateView(view, withAngle: angle, andAnimationDuration: animationDuration, usingFullCircle: false)
                }
        }
    }

// MARK: - Touches
    
    func handleTap(_ tapGesture: UITapGestureRecognizer) {
        
        UIView.animate(withDuration: 2,
                       delay: 0,
                       options: [.allowUserInteraction, .beginFromCurrentState],
                       animations: {
                        self.ball.center = tapGesture.location(in: self.view)
            })
        
    }
    
    func handleRightSwipe(_ rightSwipeGesture: UISwipeGestureRecognizer) {
        rotateView(ball, withAngle: pi, andAnimationDuration: 1, usingFullCircle: true)
    }
    
    func handleLeftSwipe(_ rightSwipeGesture: UISwipeGestureRecognizer) {
        rotateView(ball, withAngle: minus_pi, andAnimationDuration: 1, usingFullCircle: true)
    }
    
    func handleDoubleTap(_ doubleTapGesture: UITapGestureRecognizer) {
        ball.layer.removeAllAnimations()
    }
    
    func handlePinch(_ pinchGesture: UIPinchGestureRecognizer) {
        
        if pinchGesture.state == .began {
            ballScale = 1.0
        }
        
        let scale: CGFloat = 1.0 + pinchGesture.scale - ballScale
        
        ball.transform = ball.transform.scaledBy(x: scale, y: scale)
        
        ballScale = pinchGesture.scale
        
    }
    
    func handleRotation(_ rotationGesture: UIRotationGestureRecognizer) {
        
        if rotationGesture.state == .began {
            ballRotation = 0.0
        }
        
        let rotation = rotationGesture.rotation - ballRotation
        
        ball.transform = ball.transform.rotated(by: rotation)
        
        ballRotation = rotationGesture.rotation
        
    }
    
// MARK: - UIGestureRecognizerDelegate
    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer,
                           shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        return true
    }
    
}

